File: my-folder/problems/3sum/solution.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    let output = []
    nums.sort((a, b) => a - b)
    for (let i = 0; i < nums.length; i++) {
        if (nums[i] === nums[i - 1]) {
            continue
        } else {
            find(nums, nums[i], output, i + 1)
        }
    }
    return output
};

const find = (nums, target, output, left) => {
    let right = nums.length - 1
    while(left < right) {
        let curSum =  nums[left] + nums[right]
        if (curSum + target === 0) {
            output.push([target, nums[left], nums[right]])
            left++
            right--
        
            while(left > 0 && nums[left] === nums[left - 1]) {
                left++
            }
            while(right > 0 && nums[right] === nums[right + 1]) {
                right--
            }
        } else if(curSum < -target) {
            left++
        } else {
            right--
        }
        
    }
}

File: my-folder/problems/3sum/solution.py
class Solution:

    
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        output = []
        nums.sort()
        for i in range(0, len(nums) - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            else:
                self.find(nums, nums[i], i + 1, output )
        return output


    def find(self, nums, target, left, output):
        right = len(nums) - 1
        while left < right:
            sum = nums[left] + nums[right]
            if target + sum == 0:
                output.append([target, nums[left], nums[right]])
                left += 1
                right -= 1
                while left < right and nums[left] == nums[left - 1]:
                    left += 1
                while left < right and nums[right] == nums[right + 1]:
                    right -= 1
            elif sum < - target:
                left += 1
            else:
                right -= 1

                